Xamarin.Forms.Core/DispatcherExtensions.cs

```diff
@@ -0,0 +1,68 @@
+﻿using System;
+
+namespace Xamarin.Forms
+{
+	internal static class DispatcherExtensions
+	{
+		static IDispatcherProvider s_current;
+		static IDispatcher s_default;
+
+		public static IDispatcher GetDispatcher(this BindableObject bindableObject)
+		{
+			if (s_default != null)
+			{
+				// If we're already using the fallback dispatcher, keep using it
+				return s_default;
+			}
+
+			// See if the current platform has a DispatcherProvider for us
+			s_current = s_current ?? DependencyService.Get<IDispatcherProvider>();
+
+			if (s_current == null)
+			{
+				// No DispatcherProvider available, use the fallback dispatcher
+				s_default = new FallbackDispatcher();
+				return s_default;
+			}
+
+			// Use the DispatcherProvider to retrieve an appropriate dispatcher for this BindableObject
+			return s_current.GetDispatcher(bindableObject);
+		}
+
+		public static void Dispatch(this IDispatcher dispatcher, Action action)
+		{
+			if (dispatcher != null)
```

Comment:
if dispatcher and device were sharing a common interface, we could stop the braces madness, and write
```
Dispatch()
{
    IDispatcher disp = dispatcher ?? device;
    if (disp.IsInvokeRequ)
        disp.Invoke
    else
       action();
}